### PR TITLE
Dp 16437 unpublished node reminder email

### PR DESCRIPTION
**Pull request by arthurbaghdas**
**Description:**
Created new module for the logic described in Jira ticket

**Jira:** (Skip unless you are MA staff)
https://jira.mass.gov/browse/DP-16437

### CODE REVIEW

File: docroot/modules/custom/mass_unpublish_reminders/mass_unpublish_reminders.info.yml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+name: Mass.gov Node Unpublish Reminders
+type: module
+description: A module which will remind users that node will be unblished soon.
```
Comment (tom-fleming) [inline docroot/modules/custom/mass_unpublish_reminders/mass_unpublish_reminders.info.yml]:
```suggestion
description: Generates reminders for users prior to content unpublish dates.
```

File: docroot/modules/custom/mass_unpublish_reminders/mass_unpublish_reminders.module [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+          if (!empty($r->reminder_sent)) {
+            $current_timestamp = \Drupal::time()->getRequestTime();
+            $current_datetime = new DateTime();
+            $current_datetime->setTimestamp($current_timestamp);
+            $reminder_datetime = new DateTime();
+            $reminder_timestamp = $r->reminder_sent;
+            $reminder_datetime->setTimestamp($reminder_timestamp);
+            $interval = $current_datetime->diff($reminder_datetime);
+            if ($interval->days > 7) {
+              // If it is more then 7 days then the unpublish date has been
```
Comment (tom-fleming) [inline docroot/modules/custom/mass_unpublish_reminders/mass_unpublish_reminders.module]:
```suggestion
              // If interval is more than 7 days, then the unpublish date has been
```

File: docroot/modules/custom/mass_unpublish_reminders/templates/mass-reminder-mail-template.html.twig [hunk 3]
```diff
@@ -0,0 +1,8 @@
+<p>A page or alert that you authored has an unpublish date that will arrive soon.  At that time, the page or alert will be unpublished.</p>
+<p>Page: {{ page_url }}</p>
+<p>Unpublish date: {{ unpublish_date }}</p>
+<p>If you want to keep the page or alert, please review it, check it's performance, and update it if necessary.  You can then update the unpublish date to a new date.</p>
```
Comment (tom-fleming) [inline docroot/modules/custom/mass_unpublish_reminders/templates/mass-reminder-mail-template.html.twig]:
```suggestion
<p>If you want to keep the page or alert, please review it, check its performance, and update it if necessary.  You can then update the unpublish date to a new date.</p>
```

File: docroot/modules/custom/mass_unpublish_reminders/src/Plugin/QueueWorker/UnpublishReminderQueue.php [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+   */
+  public function processItem($nid) {
+    $node = Node::load($nid);
+    $author = $node->getOwner();
+    if (!empty($author)) {
+      $author_mail = $author->getEmail();
+      $params = [];
+      $cc_mails = [];
+      $unpublish_date = '';
+      $url = '';
```
Comment (tom-fleming) [inline docroot/modules/custom/mass_unpublish_reminders/src/Plugin/QueueWorker/UnpublishReminderQueue.php]: @arthurbaghdas This appears to be an unused variable. We should remove it if that is the case.

**Comment by tom-fleming** [comment]
@arthurbaghdas I made some suggested changes in the code to review. This also seems like something we want unit tests for, but might be good to get to get some guidance from @weitzman on that. I didn't see examples in other custom modules where we test custom email functionality, so not sure if it's useful here.

File: docroot/modules/custom/mass_unpublish_reminders/templates/mass-reminder-mail-template.html.twig (hunk 3 above)
Comment (arthurbaghdas) [reply docroot/modules/custom/mass_unpublish_reminders/templates/mass-reminder-mail-template.html.twig]: I don't think we should touch this. this is an approved version with changes made by Joe.

File: docroot/modules/custom/mass_unpublish_reminders/src/Plugin/QueueWorker/UnpublishReminderQueue.php (hunk 4 above)
Comment (arthurbaghdas) [reply docroot/modules/custom/mass_unpublish_reminders/src/Plugin/QueueWorker/UnpublishReminderQueue.php]: Refactored the code to remove those variables.

**Comment by arthurbaghdas** [comment]
Thank you @tom-fleming , please review again

File: docroot/modules/custom/mass_unpublish_reminders/templates/mass-reminder-mail-template.html.twig (hunk 3 above)
Comment (tom-fleming) [inline docroot/modules/custom/mass_unpublish_reminders/templates/mass-reminder-mail-template.html.twig]:
```suggestion
<p>If you want to keep the page or alert, please review it, check its performance, and update it if necessary.  You can then update the unpublish date to a new date.</p>
```

@joeg8612 Could you approve this fix?

File: docroot/modules/custom/mass_unpublish_reminders/tests/ExistingSite/UnPublishRemindersTest.php [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+use weitzman\DrupalTestTraits\Mail\MailCollectionAssertTrait;
+use weitzman\DrupalTestTraits\Mail\MailCollectionTrait;
+use weitzman\DrupalTestTraits\Entity\UserCreationTrait;
+use weitzman\DrupalTestTraits\Entity\TaxonomyCreationTrait;
+use Drupal\Tests\Traits\Core\CronRunTrait;
+
+/**
+ * Tests reminder emails functionality.
+ */
+class UnPublishRemindersTest extends ExistingSiteBase {
```
Comment (tom-fleming) [inline docroot/modules/custom/mass_unpublish_reminders/tests/ExistingSite/UnPublishRemindersTest.php]:
```suggestion
class UnpublishRemindersTest extends ExistingSiteBase {
```

Comment (arthurbaghdas) [reply docroot/modules/custom/mass_unpublish_reminders/tests/ExistingSite/UnPublishRemindersTest.php]: Changed the class name and file name.

**Comment by arthurbaghdas** [comment]
@tom-fleming ready for review.

**Comment by tom-fleming** [comment]
Looks great!

**Pull request merged**